drivers/regulator/regulator_stm32_vrefbuf.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2025 CodeWrights GmbH
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_vrefbuf
+
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/drivers/reset.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(regulator_stm32_vrefbuf, CONFIG_REGULATOR_LOG_LEVEL);
+
+struct regulator_stm32_vrefbuf_voltage {
+	uint8_t vrs;
+	uint32_t uv;
+};
+
+struct regulator_stm32_vrefbuf_data {
+	struct regulator_common_data common;
+};
+
+struct regulator_stm32_vrefbuf_config {
+	struct regulator_common_config common;
+	VREFBUF_TypeDef *base;
+	const struct stm32_pclken pclken[1];
+	const struct reset_dt_spec reset;
+	bool vrefp_output_enable;
+	const struct regulator_stm32_vrefbuf_voltage *ref_voltages;
+	uint8_t ref_voltage_count;
+};
+
+static int regulator_stm32_vrefbuf_enable(const struct device *dev)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+
+	config->base->CSR = config->base->CSR | VREFBUF_CSR_ENVR;
+
+	return 0;
+}
+
+static int regulator_stm32_vrefbuf_disable(const struct device *dev)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+
+	config->base->CSR = config->base->CSR & ~VREFBUF_CSR_ENVR;
+
+	return 0;
+}
+
+static int regulator_stm32_vrefbuf_list_voltage(const struct device *dev, unsigned int idx,
+						int32_t *volt_uv)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+
+	if (idx >= config->ref_voltage_count) {
+		return -EINVAL;
+	}
+
+	*volt_uv = config->ref_voltages[idx].uv;
+
+	return 0;
+}
+
+static unsigned int regulator_stm32_vrefbuf_count_voltages(const struct device *dev)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+
+	return config->ref_voltage_count;
+}
+
+static int regulator_stm32_vrefbuf_set_voltage(const struct device *dev, int32_t min_uv,
+					       int32_t max_uv)
+{
+	const struct regulator_stm32_vrefbuf_config *config = dev->config;
+	int32_t best_vrs = -1;
+	int32_t best_voltage = INT32_MAX;
+
+	for (uint8_t i = 0; i < config->ref_voltage_count; i++) {
+		if (!IN_RANGE(config->ref_voltages[i].uv, min_uv, max_uv) ||
+		    config->ref_voltages[i].uv >= best_voltage) {
+			continue;
+		}
+
+		best_voltage = config->ref_voltages[i].uv;
+		best_vrs = config->ref_voltages[i].vrs;
+	}
+
+	if (best_voltage == INT32_MAX) {
+		return -EINVAL;
+	}
+
+	uint8_t current_vrs = FIELD_GET(VREFBUF_CSR_VRS, config->base->CSR);
+
+	if (best_vrs == current_vrs) {
+		return 0;
+	}
+
+	config->base->CSR =
+		(config->base->CSR & ~VREFBUF_CSR_VRS) | FIELD_PREP(VREFBUF_CSR_VRS, best_vrs);
```

Comment:
Step 3: use LL functions
```suggestion
	LL_VREFBUF_SetVoltageScaling(vrs);
```